<a href="https://colab.research.google.com/github/VadymBoyko/DS-HW01/blob/main/HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [307]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import linear_model

In [308]:
df = pd.read_csv('Housing.csv')
df.head(5)

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

In [389]:
# Використання scikit-learn
X = np.array(df[['area', 'bedrooms', 'bathrooms']])
Y = np.array(df['price']/1000) # використав масштабування ціні
model = linear_model.LinearRegression()
model.fit(X, Y)
model.intercept_, model.coef_

(-173.17160763263473, array([3.78762754e-01, 4.06820034e+02, 1.38604950e+03]))

In [390]:
X = np.insert(X, 0, 1, axis=1)

In [391]:
model.predict(np.array([[10000, 4, 2]])) # погрався з прогнозами

array([8013.83505888])

In [392]:
# Аналітичне рішення
np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)

array([-1.73171608e+02,  3.78762754e-01,  4.06820034e+02,  1.38604950e+03])

Реалізуємо власний підхід

In [393]:
# напишіть функцію гіпотези лінійної регресії у векторному вигляді;
def hypothesis(x, w):
    return x @ w

In [394]:
# створіть функцію для обчислення функції втрат у векторному вигляді;
def loss(x, y, w):
    m = len(y)
    return np.sum((hypothesis(x, w) - y)**2) / (2 * m)

In [395]:
# реалізуйте один крок градієнтного спуску;
def gradient_descent_step(x, y, w, learning_rate):
    m = len(y)
    w = w - learning_rate / m * x.T @ (hypothesis(x, w) - y)
    return w

In [396]:
def fit_model(x, y, w, epoch, learning_rate, thr):
    prev_loss_value = loss(x, y, w)
    for i in range(epoch):
        w = gradient_descent_step(x, y, w, learning_rate)
        loss_value = loss(x, y, w)
        if i % 100000 == 0:
          print(prev_loss_value - loss_value)
        if prev_loss_value - loss_value < thr:
            break
        prev_loss_value = loss_value
    return w, i

In [397]:
W = np.array([1, 1, 1, 1])

In [364]:
# X_normalized = np.array(df[['area', 'bedrooms', 'bathrooms']])
# Y_normalized = np.array(df['price']/1000000)
# area_column = X_normalized[:, 0]
# area_normalized = (area_column - area_column.min()) / (area_column.max() - area_column.min())
# X_normalized = np.column_stack((area_normalized, X_normalized[:, 1:],))
# X_normalized = np.insert(X_normalized, 0, 1, axis=1)
# X_normalized

In [398]:
my_result = fit_model(X, Y, W, epoch=10000000, learning_rate=0.00000000001, thr=0.0000000001)
my_result

205.49662440060638
2.565770410001278e-05
2.5656772777438164e-05
2.565770410001278e-05
2.565723843872547e-05
2.565723843872547e-05
2.5656772777438164e-05
2.5657471269369125e-05
2.5656307116150856e-05
2.5656772777438164e-05
2.5656307116150856e-05
2.5656074285507202e-05
2.5655841454863548e-05
2.5655608624219894e-05
2.5656074285507202e-05
2.565653994679451e-05
2.5656307116150856e-05
2.5656307116150856e-05
2.5655608624219894e-05
2.5655608624219894e-05
2.5655142962932587e-05
2.565467730164528e-05
2.5655142962932587e-05
2.565537579357624e-05
2.5654910132288933e-05
2.565467730164528e-05
2.5654910132288933e-05
2.5655142962932587e-05
2.5654910132288933e-05
2.5654444471001625e-05
2.5654444471001625e-05
2.565421164035797e-05
2.5654910132288933e-05
2.565421164035797e-05
2.5653978809714317e-05
2.5654444471001625e-05
2.5653745979070663e-05
2.5653745979070663e-05
2.5653745979070663e-05
2.5653978809714317e-05
2.565351314842701e-05
2.5653745979070663e-05
2.5653745979070663e-05
2.5653280317783356e-05
2.5

(array([1.03583682, 0.85473914, 1.1359364 , 1.07657203]), 9999999)

**Висновки та аніліз результатів:**

1. для досягення результатів власного метода був змушений проводити масштабування даних, а саме ціни, для того, аби дочекатися результатів (нажаль)

2. Аналітичний метод та sklearn швидко надали однаковий результат:

***Аналітичний*** (привів з експоненціального вигляду)

-173.171608,  0.378762754,  406.820034,  1386.04950

***sklearn***

-173.17160763263473, 0.378762754, 406.820034, 1386.04950

***результати власного методу розрахунку*** прийшлось чекати дуже довго
при встановленних початкових значеннях коеф 1, 1, 1, 1 після виконная 10 000 000 ітерацій, результати не дуже схожі на ті, що були отримані першими двома методами






